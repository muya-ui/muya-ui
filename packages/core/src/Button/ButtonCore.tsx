import React, { RefObject, useMemo } from 'react';
import warning from 'warning';

import { useEventCallback } from '@muya-ui/utils';

import SpinIcon from '../Spin/SpinIcon';
import memoForwardRef from '../utils/memoForwardRef';
import useStyles from '../utils/useStyles';
import useTheme from '../utils/useTheme';
import { StyledButton, StyledLink, StyledSide } from './styled';
import { IButtonCoreWithNativeProps, IButtonNode } from './types';

const defaultStyles = {
  prefix: '',
  suffix: '',
};

const ButtonCore = memoForwardRef<IButtonNode, IButtonCoreWithNativeProps>((props, ref) => {
  // @TODO remove next breaking change
  if ('textFine' in props) {
    warning(false, '[Button]: prop `textFine` is deprecated, use `fontWeight="lighter" instead.`');
  }
  const {
    size = 'm',
    htmlType = 'button',
    disabled = false,
    component: asProp = 'button',
    children,
    textFine,
    fontWeight,
    onClick,
    type,
    width,
    loading = false,
    busy = false,
    suffixNode,
    prefixNode,
    styles,
    disableSiblingMargin,
    ...otherProps
  } = props;
  const theme = useTheme();
  const innerStyles = useStyles('button', defaultStyles, styles);
  let innerBusy = busy;

  if (loading) {
    innerBusy = true;
  }

  const childrenNode = useMemo(() => {
    let innerSuffixNode = suffixNode;
    let innerPrefixNode = prefixNode;

    if (loading && theme.components.Button.loadingPosition === 'suffix') {
      innerSuffixNode = <SpinIcon color="currentColor" />;
    } else if (loading) {
      innerPrefixNode = <SpinIcon color="currentColor" />;
    }
    return (
      <>
        {innerPrefixNode ? (
          <StyledSide {...innerStyles.prefix} theme={theme} prefixNode={innerPrefixNode}>
            {innerPrefixNode}
          </StyledSide>
        ) : null}
        {children}
        {innerSuffixNode ? (
          <StyledSide {...innerStyles.suffix} theme={theme} suffixNode={innerSuffixNode}>
            {innerSuffixNode}
          </StyledSide>
        ) : null}
      </>
    );
  }, [children, innerStyles.prefix, innerStyles.suffix, loading, prefixNode, suffixNode, theme]);

  const handleClick = useEventCallback((e: React.MouseEvent<IButtonNode, MouseEvent>) => {
    if (!innerBusy && !disabled) {
      onClick && onClick(e);
    } else {
      e.preventDefault();
    }
  });
  if (otherProps.href || asProp === 'a') {
    return (
      <StyledLink
        ref={ref as RefObject<HTMLAnchorElement>}
        role="button"
        aria-disabled={disabled}
        theme={theme}
        onClick={handleClick}
        {...otherProps}
      >
        {childrenNode}
      </StyledLink>
    );
  }

  // 只给button类型加htmlType
  const buttonType = asProp !== 'button' ? undefined : htmlType;
  return (
    <StyledButton
      ref={ref as RefObject<HTMLButtonElement>}
      as={asProp}
      type={buttonType}
      theme={theme}
      disabled={disabled}
      onClick={handleClick}
      {...otherProps}
    >
      {childrenNode}
    </StyledButton>
  );
});

export default ButtonCore;
