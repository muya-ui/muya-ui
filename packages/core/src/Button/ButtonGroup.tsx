import React from 'react';
import styled, { css } from 'styled-components';

import useTheme from '../utils/useTheme';
import { withThemeForStyled } from '../utils/withTheme';
import { siblingCss } from './mixins';
import { IButtonGroupProps, IButtonProps, IButtonType } from './types';

const BaseNode = styled.div``;
export interface IChildConfig extends Pick<IButtonProps, 'groupType' | 'offBorder'> {
  zIndex?: number;
}
export type IChildConfigMap = Record<string | number, IChildConfig>;

/* eslint-disable complexity */
/**
 * 在 Group 中需要根据 z-index 来决定谁的 border 在上面
 * @param childIndex
 * @param configMap
 * @param disabled
 */
export function getOffBorder(childIndex: number, configMap: IChildConfigMap, disabled?: boolean) {
  const { groupType, zIndex } = configMap[childIndex];
  let offBorder: IButtonProps['offBorder'];
  const prevItem = configMap[childIndex - 1];
  const nextItem = configMap[childIndex + 1];
  if (disabled && (groupType === 'head' || groupType === 'group')) {
    offBorder = 'right';
  } else if (disabled && groupType === 'tail') {
    offBorder = 'no';
  }
  // 是第一个的时候，有两种情况
  else if (groupType === 'head' && nextItem && nextItem.zIndex! < zIndex!) {
    offBorder = 'no';
  } else if (groupType === 'head') {
    offBorder = 'right';
  }
  // 是最后的一个的时候，也有两种情况
  else if (groupType === 'tail' && prevItem && prevItem.zIndex! > zIndex!) {
    offBorder = 'left';
  } else if (groupType === 'tail') {
    offBorder = 'no';
  }
  // 在中间的时候，有四种情况
  else if (groupType === 'group' && nextItem.zIndex! > zIndex! && prevItem.zIndex! > zIndex!) {
    offBorder = 'both';
  } else if (groupType === 'group' && nextItem.zIndex! < zIndex! && prevItem.zIndex! < zIndex!) {
    offBorder = 'no';
  } else if (groupType === 'group' && nextItem.zIndex! < zIndex! && prevItem.zIndex! > zIndex!) {
    offBorder = 'left';
  } else {
    offBorder = 'right';
  }
  return offBorder;
}
/* eslint-enable complexity */

const ButtonGroupWithoutStyle = React.memo(
  React.forwardRef<HTMLDivElement, IButtonGroupProps>((props, ref) => {
    const {
      type,
      size = 'm',
      htmlType,
      disabled,
      busy,
      shape,
      block,
      children,
      plain,
      ...otherProps
    } = props;
    const theme = useTheme();
    const childrenLen = React.Children.count(children);
    const notOnlyOne = childrenLen > 1;
    let resultChildren;
    const { zIndexOrder } = theme.components.Button;
    if (children) {
      const configMap: IChildConfigMap = {};
      React.Children.forEach(children, (child, childIndex) => {
        if (!React.isValidElement(child)) {
          return;
        }
        const currentType = (child.props.type as IButtonType) || type;
        let groupType: IButtonProps['groupType'];
        if (notOnlyOne && !childIndex) {
          groupType = 'head';
        } else if (notOnlyOne && childIndex === childrenLen - 1) {
          groupType = 'tail';
        } else if (notOnlyOne) {
          groupType = 'group';
        }

        configMap[childIndex] = {
          groupType,
          zIndex: zIndexOrder[currentType] || 0,
        };
      });
      resultChildren = React.Children.map(children, (child, childIndex) => {
        if (!React.isValidElement(child)) {
          return null;
        }
        const {
          disabled: childDisabled,
          busy: childBusy,
          shape: childShape,
          size: childSize,
          block: childBlock,
          type: childType,
          plain: childPlain,
          offBorder: childOffBorder,
          ...otherChildProps
        } = child.props;
        // 布尔值需要注意一下
        const currentDisabled = childDisabled !== undefined ? childDisabled : disabled;
        const offBorder: IButtonProps['offBorder'] =
          childOffBorder || getOffBorder(childIndex, configMap, currentDisabled);
        const currentType = childType || type;
        const { groupType } = configMap[childIndex];

        return React.cloneElement(child, {
          ...otherChildProps,
          disabled: currentDisabled,
          busy: childBusy !== undefined ? childBusy : busy,
          shape: childShape || shape,
          plain: childPlain !== undefined ? childPlain : plain,
          type: currentType,
          offBorder,
          groupType,
          size,
          block,
        });
      });
    }
    return (
      <BaseNode ref={ref} {...otherProps}>
        {resultChildren}
      </BaseNode>
    );
  }),
);

const ButtonGroup = styled(ButtonGroupWithoutStyle)`
  ${props => {
    const { size = 'm', block, theme, disableSiblingMargin = false } = props;
    return css`
      /* 相对定位 */
      position: relative;

      /* flex + inline */
      display: inline-flex;
      ${block && 'width: 100%;'}
      ${siblingCss(size, theme, false, disableSiblingMargin)}
    `;
  }}
`;

export default withThemeForStyled(ButtonGroup);
