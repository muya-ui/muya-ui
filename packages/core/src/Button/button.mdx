---
name: Button 按钮
menu: 组件
route: /components/button
autoDemos: true
parent: 通用
---

import { DocButtonStyle } from './demos/styled';

<DocButtonStyle />

# Button 按钮

用于触发一个即时的操作，响应用户的需求。

## 使用场景

1. 主按钮：用于主行动点，一个操作区域只能有一个主按钮。
2. 默认按钮：用于非主要行动点，可与主按钮组合使用，也可以是没有主次之分的一组按钮。
3. 文本按钮：用于最次级行动点。
4. 链接按钮：用于外链行动点。
5. 图标：
   - 添加图标：用于对按钮含义补充解释，提高按钮识别效率。
   - 仅图标：需要在较小的空间内展示尽量多的按钮。

## 按钮属性

1. 危险：警示用户该操作存在风险；
2. 警告：提醒用户该操作存在风险；
3. 禁用：行动点不可用；
4. 加载中：用于异步操作等待反馈的时候，也可以避免多次提交；
5. 蒙层按钮：用于复杂的背景中，可灵活定制样式；

## 组件

按钮共含有以下五种组件：

- `Button`: 按钮。
- `InlineButton`: 文字按钮。
- `MaskButton`: 蒙层按钮。
- `IconButton`: 图标按钮
- `ButtonGroup`: 按钮组。

{Demos}


### API

#### IButtonProps

<Props propsName="IButtonProps" omit={['textFine']} />

#### IInlineButtonProps

`IconButton` 的可用属性与 `InlineButton` 一致：

- `IconButton` === `<InlineButton type="weak" weakLevel={1}>`

<Props propsName="IInlineButtonProps" omit={['textFine']} />

#### IMaskButtonProps

<Props propsName="IMaskButtonProps" omit={['textFine']} />

### Styles

<Styles typeName="IButtonStyleKeys" componentName="button" />
