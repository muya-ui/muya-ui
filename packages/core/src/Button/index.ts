export * from './types';
export { default as ButtonGroup } from './ButtonGroup';
export { default as Button } from './Button';
export { default as InlineButton } from './InlineButton';
export { default as MaskButton } from './MaskButton';
export { default } from './Button';
