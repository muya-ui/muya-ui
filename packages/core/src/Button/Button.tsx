import React from 'react';
import styled, { css, CSSObject, ThemedStyledProps } from 'styled-components';

import { colorUtils, ITheme } from '@muya-ui/theme-light';

import { withThemeForStyled } from '../utils/withTheme';
import ButtonCore from './ButtonCore';
import { buttonSizeCSS, highlightNeutralColorButtonType, siblingCss } from './mixins';
import { IButtonNode, IButtonProps, IButtonType } from './types';

const ButtonPure = React.forwardRef<IButtonNode, IButtonProps>((props, ref) => {
  const { type, shape, block, groupType, offBorder, plain, selected, ...otherProps } = props;

  return <ButtonCore ref={ref} {...otherProps} />;
});

const buttonInGroupCss = (groupType: IButtonProps['groupType']) => {
  return css`
    ${groupType === 'head' &&
      css`
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      `}

    ${groupType === 'tail' &&
      css`
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        margin-left: -1px;
      `}
  `;
};

interface IButtonCSSConfig {
  style: CSSObject;
  hoverStyle?: CSSObject;
  clickStyle?: CSSObject;
}
export function mergeCSSConfig(a: IButtonCSSConfig, b: IButtonCSSConfig) {
  return {
    style: {
      ...a.style,
      ...b.style,
    },
    hoverStyle: {
      ...a.hoverStyle,
      ...b.hoverStyle,
    },
    clickStyle: {
      ...a.clickStyle,
      ...b.clickStyle,
    },
  };
}

const buttonBorderCSS = (
  theme: ITheme,
  plain?: boolean,
  groupType?: IButtonProps['groupType'],
  offBorder?: IButtonProps['offBorder'],
  busyOrLoading?: boolean,
) => {
  const { plainIsLight } = theme.components.Button;

  const groupCSSObj: CSSObject = {};
  // 面性的按钮在 ButtonGroup 中需要设置 border
  if ((!plain || plainIsLight) && groupType) {
    if (groupType !== 'tail') {
      groupCSSObj.borderRightWidth = '1px';
      groupCSSObj.borderRightStyle = 'solid';
    }

    if (groupType !== 'head') {
      groupCSSObj.borderLeftWidth = '1px';
      groupCSSObj.borderLeftStyle = 'solid';
    }
  }

  // 在 busy 或者 loading 时整体会有透明度，导致中间的border颜色变深
  // 需要处理一下
  if (busyOrLoading) {
    if (offBorder === 'both' || offBorder === 'right') {
      groupCSSObj.borderRightColor = 'transparent';
    }

    if (offBorder === 'both' || offBorder === 'left') {
      groupCSSObj.borderLeftColor = 'transparent';
    }
  }

  const noBorderGroupCSS = css(groupCSSObj);

  let borderCss;
  if (plain && !plainIsLight) {
    borderCss = css`
      border-style: solid;
      border-width: 1px;
    `;
  } else {
    borderCss = css`
      border-width: 0;
    `;
  }

  return css`
    ${borderCss}
    ${noBorderGroupCSS}
  `;
};

/* eslint-disable complexity */
function getOneTypeOfButton(
  theme: ITheme,
  type: IButtonType,
  interactive: boolean,
  plain?: boolean,
): IButtonCSSConfig {
  const {
    typeColor,
    defaultColor,
    typeBgImage,
    statusOpacity,
    textColor,
    lightTypeColor,
    plainIsLight,
    plainColorHighlight,
    borderColor: typeBorderColor,
    typeBgColor,
  } = theme.components.Button;
  const currentBgImage = typeBgImage[type];
  const currentColor = typeColor[type];
  const currentLightColor = lightTypeColor[type];
  const currentBgColor = typeBgColor[type];
  let currentTypeUseBgImage = !!currentBgImage;

  let borderColor;

  // 当是面性的按钮时，在 ButtonGroup 中是有 border 的
  // 需要设置 borderColor
  if (plain && plainIsLight) {
    borderColor = currentLightColor.hover;
  } else if (plain) {
    borderColor = typeBorderColor[type];
  } else {
    borderColor = currentColor.hover;
  }

  let color: string;
  if (plain) {
    color = currentColor.normal;
  } else if (textColor[type]) {
    color = textColor[type]!;
  } else {
    color = defaultColor;
  }

  let bgColor;
  if (plain && plainIsLight) {
    bgColor = currentLightColor.normal;
  } else if (plain) {
    bgColor = 'transparent';
  } else if (currentBgColor) {
    bgColor = currentBgColor.normal;
  } else {
    bgColor = currentColor.normal;
  }

  let hoverColor;
  let hoverBackgroundImage;
  let hoverBackgroundColor;
  let hoverBorderColor;

  let clickColor;
  let clickBackgroundImage;
  let clickBackgroundColor;
  let clickBorderColor;

  if (interactive) {
    if (plain && !plainIsLight) {
      hoverBackgroundColor = colorUtils.transparentize(statusOpacity.hover, color);
      clickBackgroundColor = colorUtils.transparentize(statusOpacity.click, color);
    } else if (plain) {
      hoverBackgroundColor = currentLightColor.hover;
      clickBackgroundColor = currentLightColor.click;
    } else if (currentTypeUseBgImage) {
      hoverBackgroundImage = currentBgImage!.hover;
      clickBackgroundImage = currentBgImage!.click;
    } else if (currentBgColor) {
      hoverBackgroundColor = currentBgColor.hover;
      clickBackgroundColor = currentBgColor.click;
    } else {
      hoverBackgroundColor = currentColor.hover;
      clickBackgroundColor = currentColor.click;
    }

    if (plain && plainColorHighlight && highlightNeutralColorButtonType(type)) {
      hoverColor = typeColor.primary.hover;
      clickColor = typeColor.primary.click;
      hoverBorderColor = typeColor.primary.hover;
      clickBorderColor = typeColor.primary.click;
    } else {
      hoverColor = color;
      clickColor = color;
    }
  }

  let backgroundImage;
  let backgroundColor;
  // 不是线框的按钮，同时使用背景
  if (!plain && currentTypeUseBgImage) {
    backgroundImage = currentBgImage!.normal;
  } else {
    backgroundColor = bgColor;
  }

  let busyOpacity;
  if (!interactive) {
    busyOpacity = theme.opacity.pattern.disabled;
  }

  return {
    style: {
      color,
      backgroundImage,
      backgroundColor,

      opacity: busyOpacity,
      cursor: interactive ? 'pointer' : 'not-allowed',
      borderColor,
    },
    hoverStyle: {
      color: hoverColor,
      backgroundImage: hoverBackgroundImage,
      backgroundColor: hoverBackgroundColor,

      borderColor: hoverBorderColor || borderColor,
    },
    clickStyle: {
      color: clickColor,
      backgroundImage: clickBackgroundImage,
      backgroundColor: clickBackgroundColor,

      borderColor: clickBorderColor || borderColor,
    },
  };
}

/* eslint-enable complexity */

function getDisabledButtonConfig(theme: ITheme, plain?: boolean): IButtonCSSConfig {
  const { defaultColor, plainIsLight } = theme.components.Button;
  const disabledColor = theme.colors.pattern.text.disabled;

  let color = defaultColor;
  let borderColor;
  let backgroundColor;

  // 如果是线框的
  if (plain && !plainIsLight) {
    color = disabledColor;
    borderColor = disabledColor;
    backgroundColor = 'transparent';
  } else {
    borderColor = defaultColor;
    backgroundColor = disabledColor;
  }

  return {
    style: {
      cursor: 'not-allowed',
      color,
      backgroundColor,
      borderColor,
    },
  };
}

function getButtonConfigInGroup(
  theme: ITheme,
  type: IButtonType,
  selected?: boolean,
): IButtonCSSConfig {
  const { hoverZIndex, selectedZIndex, zIndexOrder } = theme.components.Button;
  const currentZIndex = zIndexOrder[type];
  let activeZIndex;
  if (selected) {
    activeZIndex = selectedZIndex;
  } else {
    activeZIndex = hoverZIndex;
  }

  let normalZIndex;
  if (currentZIndex && selected) {
    normalZIndex = selectedZIndex;
  } else if (currentZIndex) {
    normalZIndex = currentZIndex;
  }
  return {
    style: {
      zIndex: normalZIndex,
    },
    clickStyle: {
      zIndex: activeZIndex,
    },
    hoverStyle: {
      zIndex: activeZIndex,
    },
  };
}

// 状态的样式
function buttonCSS(props: ThemedStyledProps<IButtonProps, ITheme>) {
  const {
    size = 'm',
    type = 'normal',
    plain,
    theme,
    busy,
    disabled,
    loading,
    groupType,
    offBorder,
    selected,
    disableSiblingMargin = false,
  } = props;
  const interactive = !(busy || disabled || loading);
  const baseConfig = disabled
    ? getDisabledButtonConfig(theme, plain)
    : getOneTypeOfButton(theme, type, interactive, plain);
  let config;
  if (groupType) {
    const groupConfig = getButtonConfigInGroup(theme, type, selected);
    config = mergeCSSConfig(baseConfig, groupConfig);
  } else {
    config = baseConfig;
  }

  const normalCss = css(config.style);
  let interactiveCss;
  if (interactive && config.hoverStyle && config.clickStyle) {
    interactiveCss = css`
      &:hover {
        ${css(config.hoverStyle)}
      }
      &:active {
        ${css(config.clickStyle)}
      }
    `;
  }
  const busyOrLoading = busy || loading;

  return css`
    ${normalCss}
    ${interactiveCss}
    ${buttonBorderCSS(theme, plain, groupType, offBorder, busyOrLoading)}
    ${siblingCss(size, theme, !!groupType, disableSiblingMargin)}
  `;
}

const Button = styled(ButtonPure)<IButtonProps>`
  ${props => {
    const { type, selected, groupType, plain, textFine, fontWeight } = props;
    let innerPlain = plain;
    let innerType = type;
    if (!innerType && selected === true) {
      innerType = 'primary';
    } else if (!innerType && selected === false) {
      innerType = 'secondary';
    } else if (!innerType) {
      innerType = 'normal';
    }
    if (innerType === 'normal' && plain === undefined) {
      innerPlain = true;
    }
    let innerFontWeight = fontWeight;
    if (!innerFontWeight && textFine) {
      innerFontWeight = 'lighter';
    }
    const newProps = {
      ...props,
      type: innerType,
      plain: innerPlain,
      fontWeight: innerFontWeight,
    };
    return css`
      user-select: none;
      ${buttonSizeCSS(newProps)}
      ${buttonInGroupCss(groupType)}
      ${buttonCSS(newProps)}
    `;
  }}
`;
(Button as any).__MUYA_BUTTON = true;
export default withThemeForStyled(Button);
