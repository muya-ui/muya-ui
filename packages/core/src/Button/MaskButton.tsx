import React from 'react';
import styled, { css, ThemedStyledProps } from 'styled-components';

import { colorUtils, ITheme } from '@muya-ui/theme-light';

import { withThemeForStyled } from '../utils/withTheme';
import ButtonCore from './ButtonCore';
import { buttonSizeCSS, siblingCss } from './mixins';
import { IButtonNode, IMaskButtonProps } from './types';

const MaskButtonPure = React.forwardRef<IButtonNode, IMaskButtonProps>((props, ref) => {
  const { shape, block, ...otherProps } = props;

  return <ButtonCore ref={ref} {...otherProps} />;
});

// 状态的样式
function maskButtonStatusCSS(props: ThemedStyledProps<IMaskButtonProps, ITheme>) {
  const { theme, busy, disabled, loading } = props;
  const { components, opacity } = theme;
  const { defaultColor, maskColor } = components.Button;

  let bgColor = colorUtils.transparentize(opacity.spec.s3, maskColor);
  const disabledColor = theme.colors.pattern.text.disabled;

  const interactive = !(busy || disabled || loading);
  let interactiveCSS;
  if (interactive) {
    const bgHoverColor = colorUtils.transparentize(opacity.spec.s2, maskColor);
    interactiveCSS = css`
      cursor: pointer;
      &:hover {
        background-color: ${bgHoverColor};
      }
    `;
  } else {
    interactiveCSS = css`
      cursor: not-allowed;
    `;
  }

  if (disabled) {
    bgColor = disabledColor;
  }
  return css`
    color: ${defaultColor};
    user-select: none;
    background-color: ${bgColor};
    ${!interactive && !disabled && `opacity: ${opacity.pattern.disabled};`}
    ${interactiveCSS}
  `;
}

const MaskButtonWithoutTheme = styled(MaskButtonPure)`
  ${props => {
    const { size = 'm', theme, disableSiblingMargin = false, fontWeight, textFine } = props;
    let innerFontWeight = fontWeight;
    if (!innerFontWeight && textFine) {
      innerFontWeight = 'lighter';
    }
    const newProps = {
      ...props,
      size,
      fontWeight: innerFontWeight,
    };
    return css`
      border: none;
      ${buttonSizeCSS(newProps)}
      ${maskButtonStatusCSS(newProps)}
      ${siblingCss(size, theme, false, disableSiblingMargin)}
    `;
  }}
`;

const MaskButton = withThemeForStyled(MaskButtonWithoutTheme);
(MaskButton as any).__MUYA_BUTTON = true;
export default MaskButton;
