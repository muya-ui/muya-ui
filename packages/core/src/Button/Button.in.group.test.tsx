import React from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';

test('测试 prop groupType', () => {
  const tree = renderer
    .create(
      <>
        <Button groupType="head">Test</Button>
        <Button groupType="tail">Test</Button>
        <Button groupType="group">Test</Button>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 prop offBorder', () => {
  const tree = renderer
    .create(
      <>
        <Button offBorder="right" disabled>
          Test
        </Button>
        <Button offBorder="left" disabled>
          Test
        </Button>
        <Button offBorder="both" disabled>
          Test
        </Button>
        <Button offBorder="no" disabled>
          Test
        </Button>
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
