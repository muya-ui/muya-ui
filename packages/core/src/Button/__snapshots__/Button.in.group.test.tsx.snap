// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 prop groupType 1`] = `
Array [
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  z-index: 3;
  border-style: solid;
  border-width: 1px;
}

.c1:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c1:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

<button
    className="c0 c1"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  margin-left: -1px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  z-index: 3;
  border-style: solid;
  border-width: 1px;
}

.c1:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c1:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

<button
    className="c0 c1"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 0px;
  margin-left: -1px;
  color: rgba(102,102,102,1);
  background-color: transparent;
  cursor: pointer;
  border-color: rgba(204,204,204,1);
  z-index: 3;
  border-style: solid;
  border-width: 1px;
}

.c1:hover {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.04);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

.c1:active {
  color: rgba(102,102,102,1);
  background-color: rgba(102,102,102,0.08);
  border-color: rgba(204,204,204,1);
  z-index: 9;
}

<button
    className="c0 c1"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
]
`;

exports[`测试 prop offBorder 1`] = `
Array [
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  cursor: not-allowed;
  color: rgba(221,221,221,1);
  background-color: transparent;
  border-color: rgba(221,221,221,1);
  border-style: solid;
  border-width: 1px;
}

.c2 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  cursor: not-allowed;
  color: rgba(221,221,221,1);
  background-color: transparent;
  border-color: rgba(221,221,221,1);
  border-style: solid;
  border-width: 1px;
}

.c2 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  cursor: not-allowed;
  color: rgba(221,221,221,1);
  background-color: transparent;
  border-color: rgba(221,221,221,1);
  border-style: solid;
  border-width: 1px;
}

.c2 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
  .c0 {
  outline: 0;
  margin: 0;
  font-family: inherit;
  text-transform: none;
  overflow: visible;
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 32px;
  padding: 0 13px 0 13px;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  border-radius: 2px;
  cursor: not-allowed;
  color: rgba(221,221,221,1);
  background-color: transparent;
  border-color: rgba(221,221,221,1);
  border-style: solid;
  border-width: 1px;
}

.c2 + .c1 {
  margin-left: 8px;
}

<button
    className="c0 c1 c2"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Test
  </button>,
]
`;
